Test for environment variables using the env command

gkeramidas opened the pull request:
Fixes brikis98/docker-osx-dev#19 by spawning a sub-shell and then
checking inside that shell if DOCKER_XXX environment variables
are defined, instead of checking with grep commands in only one
of the possible startup files used by the shell.

gkeramidas committed:
Test for environment variables using the env command

Fixes brikis98/docker-osx-dev#19 by spawning a sub-shell and then
checking inside that shell if DOCKER_XXX environment variables
are defined, instead of checking with grep commands in only one
of the possible startup files used by the shell.

brikis98 commented on setup.sh:
```diff
@@ -5,17 +5,14 @@
 set -e
 
+# Docker environment which we will need to install for docker-osx-dev to work.
+export   DOCKER_HOST="tcp://localhost:2375"
```
Why export this?

gkeramidas replied:
We could just set it without exporting.  Making it "readonly" makes it impossible to unset later inside env_is_defined() though.

brikis98 replied:
Ah, fair enough. Probably worth mentioning that in a comment so it's clear why it's different than the other variables in the script.

gkeramidas replied:
Done

brikis98 commented on setup.sh:
```diff
@@ -127,22 +124,50 @@ function install_vagrant_plugins {
+#   Returns 0 when VARNAME is defined for new shells, 1 otherwise.
+function env_is_defined {
+  local var="$1"
+  if test -z "${var}" ; then
```
Perhaps put this check into a function called `assert`?

gkeramidas replied:
Done. I also added a tiny backtrace() function which I've found useful in other bash scripts.

brikis98 commented on setup.sh:
```diff
@@ -127,22 +124,50 @@ function install_vagrant_plugins {
-  if grep -q "^[^#]*$DOCKER_HOST_EXPORT" "$env_file" ; then
-    log_warn "$env_file already contains \"$DOCKER_HOST_EXPORT\", will not overwrite"
+  if env_is_defined 'DOCKER_HOST' ; then
+    log_warn "${SHELL} setup already defines DOCKER_HOST will not overwrite"
```
Perhaps mention that $env_file is a good place to look for where `DOCKER_HOST` is defined?

gkeramidas replied:
Done

brikis98 commented:
Thanks for putting this together! Just a couple minor comments.

Commit:
Don't "export" DOCKER_HOST to subshells of setup.sh

Commit:
Add an assert_non_empty function which throws a backtrace on error

Instead of embedding a lengthy "if this string is empty, print this
error and exit 1", convert to an assert_non_empty() shell function
which may be useful elsewhere too.

Commit:
Point at $env_file for existing DOCKER_HOST setup

brikis98 commented on setup.sh:
```diff
@@ -55,6 +52,31 @@ function log {
+# Function to dump a 'stack trace' for failed assertions.
+function backtrace {
+  local readonly max_trace=20
+  frame=0
```
any reason `frame` isn't marked as `local`?

gkeramidas replied:
I missed this while testing the backtrace() function

Commit:
Mark 'frame' as local in backtrace()

Commit:
Explain why DOCKER_HOST is neither 'readonly' nor 'exported' by default

brikis98 commented:
Awesome, thank you!
